crypto/x509/x509_vfy.c

```diff
@@ -519,12 +520,72 @@ static int check_chain_extensions(X509_STORE_CTX *ctx)
                 ret = 1;
             break;
         }
-        if ((x->ex_flags & EXFLAG_CA) == 0
-            && x->ex_pathlen != -1
-            && (ctx->param->flags & X509_V_FLAG_X509_STRICT)) {
-            ctx->error = X509_V_ERR_INVALID_EXTENSION;
-            ret = 0;
+        /*
+         * Do the following set of checks only if strict checking is requrested
+         * and not for self-issued (including self-signed) EE (non-CA) certs
+         * because RFC 5280 does not apply to them according RFC 6818 section 2.
+         */
+        if ((ctx->param->flags & X509_V_FLAG_X509_STRICT) != 0
+            && !(i == 0 && (x->ex_flags & EXFLAG_CA) == 0
+                 && (x->ex_flags & EXFLAG_SI) != 0)) {
+            /* Check Basic Constraints according to RFC 5280 section 4.2.1.9 */
+            if (x->ex_pathlen != -1) {
+                if ((x->ex_flags & EXFLAG_CA) == 0)
+                    ctx->error = X509_V_ERR_PATHLEN_INVALID_FOR_NON_CA;
+                if ((x->ex_kusage & KU_KEY_CERT_SIGN) == 0)
+                    ctx->error = X509_V_ERR_PATHLEN_WITHOUT_KU_KEY_CERT_SIGN;
+            }
+            if ((x->ex_flags & EXFLAG_CA) != 0
+                    && (x->ex_flags & EXFLAG_BCONS) != 0
+                    && (x->ex_flags & EXFLAG_BCONS_CRITICAL) == 0)
+                ctx->error = X509_V_ERR_CA_BCONS_NOT_CRITICAL;
+            /* Check Key Usage according to RFC 5280 section 4.2.1.3 */
+            if ((x->ex_flags & EXFLAG_CA) != 0) {
+                if ((x->ex_flags & EXFLAG_KUSAGE) == 0)
+                    ctx->error = X509_V_ERR_CA_CERT_MISSING_KEY_USAGE;
```

Comment:
Should we also check for KU_KEY_CERT_SIGN?

Reply:
This is not required for general cert conformance, and it depends on the actual usage of a CA cert if keyCertSign and/or cRLSign needs to be set.

Reply:
I'm not sure how a CA cert can sign a CRL without also being able to sign a cert, but I have to admit that I don't fully understand how CRL signing certs work currently.

Reply:
See RFC 5280 seciion 5.1.1.3: 
> CAs that are also CRL issuers MAY use one private key to digitally
   sign certificates and CRLs, or MAY use separate private keys to
   digitally sign certificates and CRLs.  When separate private keys are
   employed, each of the public keys associated with these private keys
   is placed in a separate certificate, one with the keyCertSign bit set
   in the key usage extension, and one with the cRLSign bit set in the
   key usage extension (Section 4.2.1.3).

Reply:
So check for KU_KEY_CERT_SIGN|KU_CRL_SIGN?

Reply:
I guess the RFC doesn't really say anything about that.

Reply:
But I still expect KU_KEY_CERT_SIGN for i > 0

Reply:
> I guess the RFC doesn't really say anything about that.

As I wrote above, this is not part of the general (static) conformance requirements in section 4, which are meant to be implemented by `check_chain_extensions()`.
The RFC does require checking keyCertSign and cRLSign elsewhere, namely in section 6 which focuses on the dynamic cert chain checks when the actual use of certs is verified.

Reply:
The check for KU_KEY_CERT_SIGN is already implemented in `x509_signing_allowed()`, called by `internal_verify()`.
Both `check_chain_extensions()` and `internal_verify()` are called by `verify_chain()`,
so I see no need to check KU_KEY_CERT_SIGN also in `check_chain_extensions()`.